### ISSUE TITLE
[Feature] Order Stream

### ISSUE THREAD

**Issue by josojo** [gnosis/dex-contracts#88]
### Streamlining orders

The following idea outlines a concept for order collection and order cancelations. The higher level goal is to, generally, be able to consider more orders in a batch and to make order submissions cheaper.

**Order collection**
Let's assume that there is a stream of orders being submitted by the traders. These orders are sent to anchor contract and getting hashed with the rolling hash technique. However, there is not just one rolling hash per batch. Rather orders are sequentially put into _order sub-batches_, consisting of n-orders and for each order sub-batches a rolling hash is calculated. Order sub-batches are always fully filled, only at the time of closing an auction batch, the current intermediate hash of the order sub-batches is stored as well. This allows determining which orders of order sub-batches are included in an auction batch.

Every 3 minutes a batch is closed. Once a batch has closed the orders of the batch are determined. All orders submitted in the current order sub-batches plus the p-1 previous order sub-batches are being considered. In total, a maximum of p * n order can be considered for the optimization problem. 

**Solution submission**
Every solution submission will only contain the matched orders. In total, only L (L<<np) orders are allowed to be matched.
Solutions will have the following field:

- surplus,
- prices,
- order indices ( index representing the order index in the n*p orders)
- indices-hash: hash of all indices and will be calculated on the chain.
- rolling hash of matched orders
- trading volume per order

Besides the usual challenges for a solution, there needs to be a challenge determining whether the matched orders are really a subset of the given orders and whether the orders have not been matched or canceled before:

**Subset Challenge**
Given the n*p considered orders and the order indices of the solution, any client can recalculate the rolling hash of the matched orders. In case, the rolling hash of matched orders was not given correctly, the rolling hash needs to be calculated on chain. For this, the challenger will shoot 1 transaction per order sub-batches. In each transaction, the following logic is checked:
- list of matched order indices hashes to the  indices-hash
- orders of payload hash to the rolling hash of the order sub-batches
- the rolling hash of matched orders is updated, by hashing all order is in the matched order index of this order sub-batch.

After the p transaction, the calculated order hash and the hash submitted in the solution must be the same, if the solution was valid.

**Matched Challenge**
In case the order was matched before, a reference to the solution and its matched index can be verified on chain.

Reasonable numbers for the protocol would be:
n =4000, as then a sub-batch verifying transaction would cost 884*4000 gas in payload only.
p= 8
L = 1000

In this case, the order indices of the solution submission would only cost 1000 * ln(4000*8) = 15000 bits. The number of bits could be reduced by using bloom filters techniques.

**Order cancelation**
Each account maintains on chain a _cancelation hash_, encoding all the cancelations. Each trader can update the hash by providing the complete list of all indices of canceled orders. This list will be hashed and then stored as the cancelation hash. Each account has several _cancelation hashes_, each for one batch within the finalization period.

Inclusion proofs of canceled orders in solution submissions have the following logic:
- It points out the exact order index of the matched order, which was already canceled
- it needs to verify all data from the order sub-batch containing the order to recreate the exact owner of the order 
- For this owner, the canceled indices from the cancelation hash need to be resubmitted and checked that they hash to the cancelation hash of the ower
-The canceled indices must contain the matched order.

Order roll-overs are not part of the protocol. However, an order is eligible for several batches, depending on quickly the next n*p orders are being submitted.

**Fee mechanism**
compatible with basically any discussed fee model.

**Analysis**
Orderbook clearing attack:

The order book can only be cleared if n*p new orders are put into the system. With n=4000 and p=8, this would cost: 32*000*884 gas. With higher gas prices, this is already around several hundred dollars.

If clogging/ order book clearance would become a problem for the exchange, we could always increase p.

**Options**
Streams of order blobs may be running in parallel with different fee models, as for example an order space renting model. In this model, only certain traders would be allowed to submit an order to certain streams. In each batch, the last p blobs of each stream would be considered for finding the optimal solution.
